linker/src/lib.rs

```diff
@@ -1,27 +1,88 @@
+        Expression, PILFile, PilStatement, SelectedExpressions,
+    },
 };
 use number::FieldElement;
 
 /// a monolithic linker which outputs a single AIR
 pub fn link<T: FieldElement>(mut graph: PILGraph<T>) -> PILFile<T> {
-    assert_eq!(graph.objects.len(), 1, "only one machine is supported");
-    let object = graph
-        .objects
-        .remove(&Location::default().join("main"))
-        .unwrap();
-    let location = Location::default().join("main");
+    let mut queue = BTreeSet::from([Location::default().join("main")]);
 
     let mut pil = vec![];
 
-    pil.push(PilStatement::Namespace(
-        0,
-        location.to_string(),
-        Expression::Number(T::from(object.degree)),
-    ));
-
-    pil.extend(object.pil);
+    while let Some(location) = queue.pop_first() {
+        let object = graph.objects.remove(&location).unwrap();
+
+        // create a namespace for this object
+        pil.push(PilStatement::Namespace(
+            0,
+            location.to_string(),
+            Expression::Number(T::from(object.degree)),
+        ));
+        pil.extend(object.pil);
+        for link in object.links {
+            // add the target submachine to the queue
+            queue.insert(location.clone().join(link.to.loc.clone()));
+            // add the link to this namespace as a lookup
+
+            let from = link.from;
+            let to = link.to;
+
+            // the lhs is `instr_flag { inputs, outputs }`
+            let lhs = SelectedExpressions {
+                selector: Some(direct_reference(from.instr.flag())),
+                expressions: once(Expression::Number(to.function.index))
+                    .chain(
+                        from.instr
+                            .params
+                            .inputs
+                            .params
+                            .into_iter()
+                            .chain(
+                                from.instr
+                                    .params
+                                    .outputs
+                                    .into_iter()
+                                    .flat_map(|o| o.params.into_iter()),
+                            )
+                            .map(|i| {
+                                assert!(i.ty.is_none());
+                                i.name
+                            })
+                            .map(|i| direct_reference(i)),
+                    )
+                    .collect(),
+            };
+            // the rhs is `(instr_flag * latch) { inputs, outputs }`
+            // get the instruction in the submachine
+
+            let params = to.function.params;
+
+            let to_namespace = location.clone().join(to.loc.clone()).to_string();
+
+            let rhs = SelectedExpressions {
+                selector: Some(namespaced_reference(to_namespace.clone(), to.latch)),
+                expressions: once(namespaced_reference(to_namespace.clone(), "function_id"))
```

Comment:
I think "function_id" should either be a constant or part of the type, similarly to `latch` above.

Reply:
Right, I'll treat it like latch and let the user define it.